lib/GcInfo/GcInfo.cpp

```diff
@@ -0,0 +1,271 @@
+///
+/// \file
+/// \brief Implements the generation of CLR GCTables in LLILC
+///
+//===----------------------------------------------------------------------===//
+
+#include "earlyincludes.h"
+#include "GcInfo.h"
+#include "gcinfoencoder.h"
+#include "jitpch.h"
+#include "LLILCJit.h"
+#include "Target.h"
+#include "llvm/Object/StackMapParser.h"
+#include <map>
+#include <vector>
+#include <sstream>
+
+using namespace llvm;
+using namespace std;
+
+GCInfo::GCInfo(LLILCJitContext *JitContext, uint8_t *LLVMStackMapData,
+               uint8_t *CodeBlockStart) {
+  this->JitContext = JitContext;
+  this->LLVMStackMapData = LLVMStackMapData;
+  this->CodeBlockStart = CodeBlockStart;
+
+  this->Encoder = new GcInfoEncoder(JitContext->JitInfo, JitContext->MethodInfo,
+                                    new GcInfoAllocator());
+
+#if !defined(NDEBUG)
+  this->EmitLogs = JitContext->Options->LogGcInfo;
+#endif // !NDEBUG
+}
+
+void GCInfo::encodeHeader() {
+  if (EmitLogs) {
+    dbgs() << "GcTable for Function: " << JitContext->MethodName << "\n"
+           << "  Size: " << JitContext->HotCodeSize << "\n";
+  }
+
+  // TODO: Set Code Length accurately.
+  // https://github.com/dotnet/llilc/issues/679
+  // JitContext->HotCodeSize is the size of the allocated code block.
+  // It is not the actual length of the current function's code.
+  Encoder->SetCodeLength(JitContext->HotCodeSize);
+  Encoder->SetSizeOfStackOutgoingAndScratchArea(0);
+}
+
+void GCInfo::encodeLiveness() {
+  if (LLVMStackMapData == nullptr) {
+    return;
+  }
+
+  ArrayRef<uint8_t> StackMapContentsArray(
+      reinterpret_cast<const uint8_t *>(LLVMStackMapData),
+      JitContext->StackMapSize);
+
+#if defined(BIGENDIAN)
+  typedef StackMapV1Parser<support::big> StackMapParserType;
+#else
+  typedef StackMapV1Parser<support::little> StackMapParserType;
+#endif
+  StackMapParserType StackMapParser(StackMapContentsArray);
+
+  assert(StackMapParser.getNumFunctions() == 1 &&
+         "Expect only one function with GcInfo in the module");
+
+  // The InstructionOffsets for Call-sites are with respect to:
+  // (1) FunctionEntry in LLVM's StackMap
+  // (2) CodeBlockStart in CoreCLR's GcTable
+  //
+  // There is typically a difference between the two even in the JIT case
+  // (where we emit one function per module) because of some additional
+  // code like the gc.statepoint_poll() method.
+
+  uint8_t *FunctionEntry =
+      (uint8_t *)StackMapParser.getFunction(0).getFunctionAddress();
+
+  size_t OffsetCorrection = FunctionEntry - CodeBlockStart;
+
+  // Loop over LLVM StackMap records to:
+  // 1) Note CallSites (safepoints)
+  // 2) Assign Slot-IDs to each unique gc-pointer location (slot)
+  // 3) Record liveness (birth/death) of slots per call-site.
+
+  size_t NumCallSites = StackMapParser.getNumRecords();
+  unsigned *CallSites = new unsigned[NumCallSites];
+  BYTE *CallSiteSizes = new BYTE[NumCallSites];
```

Comment:
Instead of always allocating arrays on the heap, you might want to use `llvm::SmallVector` with a size of 16 or so. Values of that type will use the stack until it runs out of room, and will deallocate any storage when they go out of scope. FWIW, these values are currently being leaked (they're never `delete[]`d).